processor.go

```diff
@@ -80,36 +83,45 @@ func (pro *Processor) OnProposal(proposal *message.Proposal) error {
 	// 1) try to confirm the parent vertex of the proposal
 	err := pro.confirmParent(proposal)
 	if err != nil {
+		// this error contains information about the reason for non-confirmation and is eventually handled in the caller scope
 		return rich.Errorf("could not confirm parent: %w", err)
 	}
 
+	// parent is confirmed -> move to applying the candidate message
+
 	// 2) try to apply the candidate vertex of the proposal
 	err = pro.applyCandidate(proposal)
 	if err != nil {
+		// this error contains information about the reason for not applying candidate and is eventually handled in the caller scope
```

Comment:
Again, not relevant. What does applying the candidate mean?

Reply:
After the verification pass successfully, it tries to extend the graph the peer knows of with the new vertex - candidate
and check if the received proposal is in its cache records.
If the new candidate vertex is included in the records - this means that there is a double candidate that has already been received.
Then if the proposal is made for a second time by a leader (which is illegal) - send error;
In my understanding the step with extending the graph is before the step with trying to cache the proposal (which might return an error) because it is assumed that if the proposal candidate is valid - it should be added to the graph by all honest peers
AND
it makes sense to check if it was added to graph already - because every other peer would add a valid candidate to the graph anyway.